src/main/java/epam/task/gymbootdb/repository/TrainingRepository.java

```diff
@@ -0,0 +1,35 @@
+package epam.task.gymbootdb.repository;
+
+import epam.task.gymbootdb.entity.Training;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+
+import java.time.LocalDate;
+import java.util.List;
+
+public interface TrainingRepository extends JpaRepository<Training, Long> {
+
+    @Query("SELECT t FROM Training t WHERE t.trainee.user.username = :traineeUsername AND " +
+            "(:fromDate IS NULL OR t.date >= :fromDate) AND " +
+            "(:toDate IS NULL OR t.date <= :toDate) AND " +
+            "(:trainerUsername IS NULL OR t.trainer.user.username = :trainerUsername) AND " +
+            "(:trainingTypeName IS NULL OR t.trainingType.name = :trainingTypeName)")
+    List<Training> findTraineeTrainingsByOptionalParams(
+            @Param("traineeUsername") String traineeUsername,
```

Comment:
@Param annotation is redundant if the method argument is the same as a query parameter

Without annotation, code will be more readable